### ISSUE TITLE
Compilation Error 

### ISSUE THREAD

**Issue by joelsherrill** [Rust-GCC/gccrs#1376]
Attempting to compile gccrs to target arm-rtems6 using a native gcc built from the gcc head. The last gccrs commit was:

`commit a273dbd971d337000155fe56a1d7cd78868caa9a (HEAD -> master, origin/staging, origin/master, origin/HEAD)
Merge: 3f325921597 3ccf43082d2
Author: bors[bot] <26634292+bors[bot]@users.noreply.github.com>
Date:   Fri Jul 8 18:19:34 2022 +0000

    Merge #1374
    
    1374: remove unused go code r=philberty a=philberty
    
    This function is not used and we don't want to disable optimizations
    randomly like this.
    
    Co-authored-by: Philip Herron <philip.herron@embecosm.com>
`
and compiling with gcc (not gccrs) built from gcc master:

` $ gcc --version
gcc (GCC) 13.0.0 20220710 (experimental)
Copyright (C) 2022 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
`
I got this error:

```g++  -fno-PIE -c  -DIN_GCC_FRONTEND -g -O2     -DIN_GCC  -DCROSS_DIRECTORY_STRUCTURE   -fno-exceptions -fno-rtti -fasynchronous-unwind-tables -W -Wall -Wno-narrowing -Wwrite-strings -Wcast-qual -Wno-error=format-diag -Wmissing-format-attribute -Woverloaded-virtual -pedantic -Wno-long-long -Wno-variadic-macros -Wno-overlength-strings -fno-common  -DHAVE_CONFIG_H -I. -Irust -I../../gccrs/gcc -I../../gccrs/gcc/rust -I../../gccrs/gcc/../include -I../../gccrs/gcc/../libcpp/include -I../../gccrs/gcc/../libcody -I/home/joel/test-gcc/b-arm-rtems6-gcc/./gmp -I/home/joel/test-gcc/gccrs/gmp -I/home/joel/test-gcc/b-arm-rtems6-gcc/./mpfr/src -I/home/joel/test-gcc/gccrs/mpfr/src -I/home/joel/test-gcc/gccrs/mpc/src  -I../../gccrs/gcc/../libdecnumber -I../../gccrs/gcc/../libdecnumber/dpd -I../libdecnumber -I../../gccrs/gcc/../libbacktrace -I/home/joel/test-gcc/b-arm-rtems6-gcc/./isl/include -I/home/joel/test-gcc/gccrs/isl/include  -o rust/rust-cfg-parser.o -MT rust/rust-cfg-parser.o -MMD -MP -MF rust/.deps/rust-cfg-parser.TPo -std=c++11 -Wno-unused-parameter -Werror=overloaded-virtual -I ../../gccrs/gcc/rust -I ../../gccrs/gcc/rust/lex -I ../../gccrs/gcc/rust/parse -I ../../gccrs/gcc/rust/ast -I ../../gccrs/gcc/rust/analysis -I ../../gccrs/gcc/rust/backend -I ../../gccrs/gcc/rust/expand -I ../../gccrs/gcc/rust/hir/tree -I ../../gccrs/gcc/rust/hir -I ../../gccrs/gcc/rust/resolve -I ../../gccrs/gcc/rust/util -I ../../gccrs/gcc/rust/typecheck -I ../../gccrs/gcc/rust/privacy -I ../../gccrs/gcc/rust/lint -I ../../gccrs/gcc/rust/util ../../gccrs/gcc/rust/parse/rust-cfg-parser.cc
In file included from ../../gccrs/gcc/rust/ast/rust-ast-full.h:22,
                 from ../../gccrs/gcc/rust/parse/rust-parse.h:21,
                 from ../../gccrs/gcc/rust/parse/rust-cfg-parser.cc:3:
../../gccrs/gcc/rust/ast/rust-ast.h:120:16: error: ‘virtual void Rust::AST::MacroMatch::accept_vis(Rust::AST::ASTVisitor&)’ was hidden [-Werror=overloaded-virtual=]
  120 |   virtual void accept_vis (ASTVisitor &vis) = 0;
      |                ^~~~~~~~~~
../../gccrs/gcc/rust/ast/rust-ast.h:223:8: note:   by ‘virtual void Rust::AST::Token::accept_vis(Rust::AST::ASTVisitor&)’
  223 |   void accept_vis (ASTVisitor &vis) override;
      |        ^~~~~~~~~~
../../gccrs/gcc/rust/ast/rust-ast.h:111:23: error: ‘virtual std::string Rust::AST::MacroMatch::as_string() const’ was hidden [-Werror=overloaded-virtual=]
  111 |   virtual std::string as_string () const = 0;
      |                       ^~~~~~~~~
../../gccrs/gcc/rust/ast/rust-ast.h:220:15: note:   by ‘virtual std::string Rust::AST::Token::as_string() const’
  220 |   std::string as_string () const override;
      |               ^~~~~~~~~
../../gccrs/gcc/rust/ast/rust-ast.h:593:16: error: ‘virtual void Rust::AST::AttrInput::accept_vis(Rust::AST::ASTVisitor&)’ was hidden [-Werror=overloaded-virtual=]
  593 |   virtual void accept_vis (ASTVisitor &vis) = 0;
      |                ^~~~~~~~~~
../../gccrs/gcc/rust/ast/rust-ast.h:798:8: note:   by ‘virtual void Rust::AST::DelimTokenTree::accept_vis(Rust::AST::ASTVisitor&)’
  798 |   void accept_vis (ASTVisitor &vis) override;
      |        ^~~~~~~~~~
../../gccrs/gcc/rust/ast/rust-ast.h:591:23: error: ‘virtual std::string Rust::AST::AttrInput::as_string() const’ was hidden [-Werror=overloaded-virtual=]
  591 |   virtual std::string as_string () const = 0;
      |                       ^~~~~~~~~
../../gccrs/gcc/rust/ast/rust-ast.h:796:15: note:   by ‘virtual std::string Rust::AST::DelimTokenTree::as_string() const’
  796 |   std::string as_string () const override;
      |               ^~~~~~~~~
```

I expected to see this happen: *explanation*

Instead, this happened: *explanation*

### Meta

- What version of Rust GCC were you using, git sha if possible.

**Comment by CohenArthur** [comment]
Interesting. I wonder why these C++ errors didn't pop up on all our other builders. It might be due to using the latest version of gcc? I wonder if that happens on x86 as well

**Comment by Zopolis4** [comment]
I'm using gcc 12.1, had no build fails, although I was disabling bootstrap (although most of a bootstrap build completed before that). There arent any changes in gcc that I can see https://gcc.gnu.org/gcc-13/changes.html that would have affected this code area.

**Comment by philberty** [comment]
It looks as though there has been a patch submitted to GCC master which might fix this for you:

```
commit 4c6567b4090d9363f90b5c68e238e7d0964b6e89
Author: Jason Merrill <jason@redhat.com>
Date:   Sun Jul 24 23:26:59 2022 -0400

    c++: -Woverloaded-virtual false positive [PR87729]
```

We usually build gccrs with our distro compiler at the moment.

**Comment by philberty** [comment]
@CohenArthur is going to check this for us :D

**Comment by CohenArthur** [comment]
Hi @joelsherrill,

I've tried building two `gcc`s, before and after [this commit](https://gcc.gnu.org/git/?p=gcc.git;a=commit;h=4c6567b4090d9363f90b5c68e238e7d0964b6e89) which fixes some overzealous virtual methods warnings.

The "before" build failed with the same error as yours. The "after" one however completed without an issue. I did full bootstrap builds for both versions of `gcc`, to get as close to your scenario as possible. Hope it helps :)
